frontend/packages/dev-console/src/components/topology/data-transforms/topology-model.ts

```diff
@@ -0,0 +1,166 @@
+import { EdgeModel, Model, NodeModel, NodeShape, createAggregateEdges } from '@console/topology';
+import { TopologyFilters } from '../filters/filter-utils';
+import { TopologyDataModel, TopologyDataObject, Node, Group } from '../topology-types';
+import {
+  TYPE_APPLICATION_GROUP,
+  TYPE_HELM_RELEASE,
+  TYPE_AGGREGATE_EDGE,
+  TYPE_OPERATOR_BACKED_SERVICE,
+  NODE_WIDTH,
+  NODE_HEIGHT,
+  NODE_PADDING,
+  GROUP_WIDTH,
+  GROUP_HEIGHT,
+  GROUP_PADDING,
+} from '../const';
+import { ALL_APPLICATIONS_KEY } from '@console/shared/src';
+import {
+  getNodeModel as getKnativeNodeModel,
+  getGroupModel as getKnativeGroupModel,
+  getEdgeModel as getKnativeEdgeModel,
+} from '@console/knative-plugin/src/topology/knative-topology-model';
+
+export const dataObjectFromModel = (node: Node | Group): TopologyDataObject => {
+  return {
+    id: node.id,
+    name: node.name,
+    type: node.type,
+    resources: null,
+    operatorBackedService: false,
+    data: null,
+  };
+};
+
+const getApplicationGroupForNode = (node: Node, groups: NodeModel[]): NodeModel => {
+  const group = groups.find((g) => g.children.includes(node.id));
+  if (!group) {
+    return null;
+  }
+  if (group.type === TYPE_APPLICATION_GROUP) {
+    return group;
+  }
+  return getApplicationGroupForNode(group, groups);
+};
+
+export const topologyModelFromDataModel = (
+  dataModel: TopologyDataModel,
+  application: string = ALL_APPLICATIONS_KEY,
+  filters?: TopologyFilters,
+): Model => {
+  const groupNodes: NodeModel[] = dataModel.graph.groups.map((d) => {
```

Comment:
We should look to remove the intermediate data model and handle groups as nodes in the future. Probably too much to do as part of this refactor.